infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -596,9 +592,24 @@ public void processTopologyChangeNotification(DiscoveryServiceEvent event) {
      * false, otherwise
      */
     private boolean clusterRoleChanged(TopologyDescriptor discoveredTopology) {
-        return topologyDescriptor.getLocalClusterDescriptor() != null ?
-                topologyDescriptor.getLocalClusterDescriptor().getRole() !=
-                discoveredTopology.getLocalClusterDescriptor().getRole() : false;
+        // find the remote clusters which have been removed from remoteSourceClusters and added to remoteSinkClusters in
+        // the new topology.
+        Set<String> remoteSourceRemoved = Sets.difference(topologyDescriptor.getRemoteSourceClusters().keySet(),
+                discoveredTopology.getRemoteSourceClusters().keySet());
+        Set<String> remoteSinkAdded = Sets.difference(discoveredTopology.getRemoteSinkClusters().keySet(),
+                topologyDescriptor.getRemoteSinkClusters().keySet());
+        if (!Sets.intersection(remoteSinkAdded, remoteSourceRemoved).isEmpty()) {
+            return true;
+        }
+
+        // find the remote clusters which have been removed from remoteSinkClusters and added to remoteSourceClusters in
+        // the new topology
+        Set<String> remoteSinkRemoved = Sets.difference(topologyDescriptor.getRemoteSinkClusters().keySet(),
+                discoveredTopology.getRemoteSinkClusters().keySet());
+        Set<String> remoteSourceAdded = Sets.difference(discoveredTopology.getRemoteSourceClusters().keySet(),
+                topologyDescriptor.getRemoteSourceClusters().keySet());
+
+        return !Sets.intersection(remoteSourceAdded, remoteSinkRemoved).isEmpty();
```

Comment:
Sorry I didn't understand this, why are we checking the change of remote Sink/Source? If an existed Sink(Source) becomes a new Source(Sink), we do `onClusterRoleChange()` for local cluster?

Reply:
Discussed offline that this is valid to cover active standby role switchover case